Multi sensitive mpgd clusters

ybedfer opened the pull request:
### Briefly, what does this PR introduce?
Implement Clusters into `MultiSensitiveMPGD`, extending  "PR#https://github.com/eic/EICrecon/pull/2177".

### What kind of change does this PR introduce?
- [ *] New feature

### Please check if this PR fulfills the following:
- [* ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
Breaking changes if and only if MPGD XMLs from `Multi-SensitiveVolume-MPGD`  (see "https://github.com/eic/epic/pull/1014") are loaded or "-PMPGD:SiFactoryPattern" option is specified and set != 3.

### Does this PR change default behavior?
No (_w.r.t. PR #2177_).

Commit:
MPGDTrackerDigi: Multi-SensitiveVolume solution for the 2D-strip readout.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTrackerDigi.cc: Fix compilation warnings.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTrackerDigi.cc: Fix codespell issues.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTrackerDigi.cc: Getting rid of "dangling-reference" compilation warning.

Commit:
Merge branch 'MultiSensitiveMPGD' of github.com:eic/EICrecon into MultiSensitiveMPGD

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTrackerDigi.cc: Fix compilation warning.

Commit:
Merge branch 'MultiSensitiveMPGD' of github.com:eic/EICrecon into MultiSensitiveMPGD

veprbl committed:
`[=, this]` -> `[&]`

```
 /home/runner/work/EICrecon/EICrecon/src/algorithms/digi/MPGDTrackerDigi.cc:156:21: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
  156 |   m_stripRank = [=, this](CellID vID) {
      |                   ~~^~~~
1 error generated.
```

wdconinc committed:
MPGDTrackerDigi: Multi-SensitiveVolume solution for the 2D-strip read… (fix: iwyu) (#2180)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/19181728533.
Please merge this PR into the branch `MultiSensitiveMPGD`
to resolve failures in PR #2177.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD

Commit:
MPGDTrackerDigi.cc: Implement Clang-Tidy suggestions.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc by Clang-Tidy.

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc by Clang-Tidy.

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
TrackerMeasurementFromHits: Rotate the cov. matrix of OuterMPGDBarrel.
This, when it happens to use UV segmentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
Update src/algorithms/tracking/TrackerMeasurementFromHits.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD

wdconinc committed:
MPGDTrackerDigi: Multi-SensitiveVolume solution for the 2D-strip read… (fix: iwyu) (#2189)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/19430351038.
Please merge this PR into the branch `MultiSensitiveMPGD`
to resolve failures in PR #2177.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Clang-Tidy.

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Clang-Tidy

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTrackerDigi: Remove debugging leftovers. Improve handling of inconsistences.

Commit:
MPGDTrackerDigi: VolumeID mask from IDDescriptor; debugging leftovers removed.
- The VolumeID mask was previously built-in.
- Debugging leftovers were not removed in previous commit (#bc26b254), by mistake.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTrackerDigi: Fix spelling.

veprbl committed:
rm unused variable

Commit:
Merge branch 'MultiSensitiveMPGD' of github.com:eic/EICrecon into MultiSensitiveMPGD

Commit:
Merge branch 'MultiSensitiveMPGD' of github.com:eic/EICrecon into MultiSensitiveMPGD

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
MPGDTrackerDigi: Clang-Tidy stuff...

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
MPGDTrackerDigi: Multi-SensitiveVolume solution for the 2D-strip read… (fix: iwyu) (#2196)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/19527585648.
Please merge this PR into the branch `MultiSensitiveMPGD`
to resolve failures in PR #2177.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
MPGDTrackerDigi: More modular design. Check Segmentation of MPGDs is appropriate.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGDTRackerDigi: Fix spelling mistake.

wdconinc committed:
MPGDTrackerDigi: Multi-SensitiveVolume solution for the 2D-strip read… (fix: iwyu) (#2199)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/19599402189.
Please merge this PR into the branch `MultiSensitiveMPGD`
to resolve failures in PR #2177.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
MPGD Barrel: Default digitization factory determined from XML file (and cosmetics...)
- Constant "<detector>_2DStrip" is expected to be defined in XML file.
- If successfully queried and its value turns out !=0, "MPGDTrackerDigi" becomes default.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGD Barrel: Fix Clang-Tidy warning in "MPGD.cc".

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wdconinc committed:
MPGDTrackerDigi: Multi-SensitiveVolume solution for the 2D-strip read… (fix: iwyu) (#2204)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/19710271095.
Please merge this PR into the branch `MultiSensitiveMPGD`
to resolve failures in PR #2177.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD

Commit:
trigger GitHub actions

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc: Restricted 'catch' condition.

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ybedfer committed:
Update src/algorithms/tracking/TrackerMeasurementFromHits.cc: Restrict 'catch' condition.

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ybedfer committed:
Update src/detectors/MPGD/MPGD.cc: Restrict 'catch' condition.

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc: Fix bug (thanks to D. Kalinkin)..

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc: Bug fix (thanks to D.Kalinkin).

Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

Commit:
MPGD: Various fixes to conform to PR-reviewer requests.

Commit:
MPGD Barrel: "MPGDTrackerDigi", "OuterMPGDBarrel" in "TrackerMeasurementFromHits", "MPGD" plugin.
- "MPGDtrackerDigi":
+ JException -> std::runtime_error.
+ FieldIDs no longer hardcoded.
+ Status code values defined as named constants.
- "OuterMPGDBarrel" in "TrackerMeasurementFromHits":
+ "gridAngle" parameter no longer fixed = 45°.
+ Parsing of Segmentation: reshuffled.
- "MPGD" plugin:
+ Bug fix in printout.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD

Commit:
MPGDTrackerDigi: Some cleaning...

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD

Commit:
MPGD: Bug fix in "TrackerMeasurementFromHits".

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
MPGD: Implement Clusters into MultiSensitiveMPGD.
Adding "MPGDHitReconstruction" and modifying "MPGDTrackerDigi" and MPGD plugin.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer commented:
This elaborates on PR #2177 (see "https://github.com/eic/EICrecon/pull/2177"), implementing Clusters.
 - Implementing them at the digitization level (_operation I call "clusterization"_).
 - Processing the clusters to produce `TrackerHits`  (_"clustering"_) in a newly added `MPGDHitReconstruction` class.

At present, the clusters are simplistic: 2-hit clusters.

It is enabled by loading dedicated `_2DStrip` MPGD XMLs, with a special `<detector_name>__2DStrip` constant defined, just as PR #2177 (_and also if "-PMPGD:SiFactoryPattern" option is specified and set != 3_). For the clusterization to be meaningful, a new version of the XMLs is needed, which requires new `epic` PR, elaborating on PR #1014.

Checks done:
 - Hit-to-Track association: 
[ClusterTrackAssociation.pdf](https://github.com/user-attachments/files/24244256/ClusterTrackAssociation.pdf)
 - Residuals RecHit-SimHit: 
[cCyMBaLRes_mu_q0.pdf](https://github.com/user-attachments/files/24244268/cCyMBaLRes_mu_q0.pdf)
[cOuterRes_mu_q0.pdf](https://github.com/user-attachments/files/24244270/cOuterRes_mu_q0.pdf)
 (SimHits here being coalesced ones, following a procedure mimicking what's done in `MPGDTrackerDigi`. )

Commit:
MPGDHitReconstruction: Bug fixes: reordering, channel# = 0xffff.
- Reordering: properly handle events w/ multiple hits per module.
- channel# = 0xffff: work around the fact that 0xffff is not understood as =-1.

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD_Clusters

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer commented:
Two fixes committed in #b71832ad.

New residuals RecHit-SimHit are now more in line w/ expectations, _viz._ Std Dev = 150 µm :
[cCyMBaLRes_mu_Q0.pdf](https://github.com/user-attachments/files/24366972/cCyMBaLRes_mu_Q0.pdf)
[cOuterRes_mu_Q0.pdf](https://github.com/user-attachments/files/24366975/cOuterRes_mu_Q0.pdf)

ybedfer committed:
MPGD: Fix compilation warnings (HitReconstruction), debugging printout (Digi).

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
MPGD plugin and digitization:
- Digitization: better handling reEntrance through lower wall.
 => This removes many (if not all) of the "inconsistency" messages showing up there.
  Also fix little bug affecting "hit extension".
- Plugin: fix printout mistake.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
MPGD digitizarion: Fix debugging printout (using indices instead of pointers).

ybedfer committed:
More robust COALESCENCE + Better handling of used hits + Simplifications.

ybedfer committed:
MPGD Digitization and Hit Reconstruction:
Digitization: Intermixed hits. Tolerance. ReEntrance. Outside hits. Clusterization. Cosmetics.
- Intermixed hits (i.e. subHits of same ancestry not in sequence):
+ There's no provision any longer for this case:
++ Corresponding piece of software had problems w/ looping hits.
++ I haven't found any instance of intermixed sequence in any of DD4hep data.
- Tolerance: is relaxed for low energy stuff.
- Outside hits (i.e. outside of sensitive volume):
+ They happen in the Tube geometry case.
+ They used to trigger inconsistencies. Now fixed.
- ReEntrance: No longer extend subHits on a reEntrance path.
- Clusterization: Bug affecting clusters combining >=0 and <0 hits fixed. This,
 both in clusterization (Digitization) and clustering (Hit Reconstruction).
- Cosmetics: No longer use "JException", remove debugging leftovers.

ybedfer committed:
MPGD Digitization and Hit Reconstruction: Cylinder edge. Links. Cosmetics.
- Cylinder edge: fix mathematics.
- Links: write *Links for all *Associations when EDM4EIC >= 8.7.0 (#2372).
- Cosmetics: include-what-you-use fixes, string_view, ...

ybedfer committed:
MPGD plugin: write *Links for all *Associations when EDM4EIC >= 8.7.0 (#2372).

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD_Clusters

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
MPGD Digitization and plugin: BELOW-THRESHOLD RawHits and cleaning.
- BELOW-THRESHOLD RawHits: while applying unique thresholds to the 2 cluster components.
 Also fix little bug to get correct charge when accumulating hits.
- "MPGD.cc": Cleaning layout.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer commented:
[pre-commit.ci] auto fixes (https://github.com/eic/EICrecon/pull/2264/commits/e10c282c8d67e411d9e37ab7901f4dd570a72135) keeps changing the layout of `MPGD.cc` in a way:
  - that is ugly,
  - and that is anyway not consistent.

ybedfer committed:
MPGD digitization: Handling of edge-crossing for OuterBarrel. Cosmetics...
- A bug was affecting subroutines "bTraversing" and "bExtension".
- It had very little impact, because:
+ edge crosssings are very rare,
+ their mishandling in "bTraversing" is most of the time rectified later on in the code,
+ the impact of edge effects is small in itself.

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD_Clusters

ybedfer committed:
MPGD digitization: undo part of commit #c59f64d9.

veprbl committed:
Merge branch 'main' into MultiSensitiveMPGD_Clusters

ybedfer committed:
MPGD Hit reconstruction: fix compilation warnings.

ybedfer committed:
MPGD Hit Reconstruction: fix compilation warning.

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
MPGD digitization: Fix Clang-Tidy issues.

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer committed:
Update src/algorithms/digi/MPGDTrackerDigi.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
MPGD digitization: Fix Clang-Tidy issues.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
MPGD digitization: Undo part of previous commit.

ybedfer committed:
Merge branch 'MultiSensitiveMPGD_Clusters' of github.com:eic/EICrecon into MultiSensitiveMPGD_Clusters

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ybedfer committed:
MPGD plugin and digitization: "gsl::at()" instead of array subscripts.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

epic-capybara committed:
Multi sensitive mpgd clusters (fix: iwyu) (#2585)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/23588765842.
Please merge this PR into the branch `MultiSensitiveMPGD_Clusters`
to resolve failures in PR #2264.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ybedfer commented:
> Would you be able to contribute your changes back to the version in git?

There's a problem: in the OuterBarrel case, as I said supra, I transform the RecHit position from Master to Local. I think I cannot avoid to do this. I am not 100% sure of the fact, though. And the parameters of the transform are built-in. Meaning that the macro would cease to work when some of the parameters of the XML (IDDescriptor, position along Z) are changed. A robust modified version would require that some kind of info (you mentioned yourself a check sum) be stored in the podio file allowing to retrieve those parameters.

veprbl committed:
Merge branch 'main' into MultiSensitiveMPGD_Clusters

veprbl commented on src/algorithms/digi/MPGDTrackerDigi.cc:
```diff
@@ -294,109 +303,172 @@ void MPGDTrackerDigi::process(const MPGDTrackerDigi::Input& input,
+      }
+      // ***** DEBUGGING INFO
+      if (level() >= algorithms::LogLevel::kDebug) {
+        std::string sCellID = (pn != 0) ? "cellIDn" : "cellIDp";
```
Inconsistent with `(pn != 0) ? 'p' : 'n'` just above.

ybedfer replied:
Fixed.

veprbl commented on src/detectors/MPGD/MPGD.cc:
```diff
@@ -148,30 +179,51 @@ void InitPlugin(JApplication* app) {
         },
         app));
+  } else {
+    MPGDHitReconstructionConfig digi_cfg;
```
```suggestion
    MPGDHitReconstructionConfig reco_cfg;
```

ybedfer replied:
Done.

ybedfer committed:
Merge branch 'main' into MultiSensitiveMPGD_Clusters

veprbl committed:
add tracking_MPGDHitReconstruction unit test

veprbl reviewed (commented):
Couple thing I spotted so far.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl commented on src/tests/algorithms_test/algorithmsInit.cc:
```diff
@@ -61,6 +68,99 @@ class algorithmsInitListener : public Catch::EventListenerBase {
     detector->add(id_desc_Silicon);
     detector->add(readoutSilicon);
 
+    // Mock MPGD readout with full geometry registered in the VolumeManager.
```
What you think about this approach, @wdconinc? Or we just load full detector?

wdconinc replied:
At some point maybe we need to add a local test geometry here, in xml with standard DDDetector plugins?

veprbl replied:
My original idea for these was to implement mocks instead of using proper DD4hep. If we commit to real geometry further, it might make sense to just trivial setups to epic. We could also advertise them for beam tests.

veprbl committed:
fix for older DD4hep

epic-capybara committed:
Multi sensitive mpgd clusters (fix: iwyu) (#2672)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/26429944306.
Please merge this PR into the branch `MultiSensitiveMPGD_Clusters`
to resolve failures in PR #2264.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
remove some fluff

ybedfer committed:
MPGD plugin and digitization: implement modifs suggested by Dmitry Kalinkin.

veprbl committed:
add unit test for MPGDTrackerDigi

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
trick spellchecker

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

epic-capybara committed:
Multi sensitive mpgd clusters (fix: iwyu) (#2673)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/26470145353.
Please merge this PR into the branch `MultiSensitiveMPGD_Clusters`
to resolve failures in PR #2264.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
update credits

veprbl reviewed (approved):
I have no more comments on this.

Tests were vibe coded. I did not push it to look for edge cases, but hopefully it will at least help with maintainability.
